### DISCUSSION TITLE
attrs vs Pydantic

### DISCUSSION THREAD

**Discussion by dmsfabiano** [fastapi/fastapi#5121] [Questions]
Hi all!

Very broad question, I am a big fan of both [attrs ](https://github.com/python-attrs/attrs)and [pydantic](https://github.com/samuelcolvin/pydanticcc). Recently, I have been definitely more into attrs, knowing the fact that attrs yields to better performance I am curious on why it pydantic was chosen over attrs (or even dataclasses) and if there are plans to include compatibility for it?

**Comment by zoliknemet** [comment]
There are few cool things coming in pydantic v2, look at this https://pydantic-docs.helpmanual.io/blog/pydantic-v2/

**Comment by SoundsSerious** [comment]
I agree with @dmsfabiano that attrs yields better performance, and can be far more customized than Pydantic, due to its class-decorator. Class decorators enable major performance benefits when initializing custom features, where instance based functionality like PyDantic must be re-performed every __init__.

One way to enable multiple styles of class without a significant burden in development would be for FASTAPI to develop its own `class-decorator` that converts various base classes to appropriate format.

```
@fastapi.define_class
class Attrs:
      field: str = attrs.field()

@fastapi.define_class
class Current(BaseModel):
      field: str
```

I dont really know enough about FASTAPI's class use to establish what should happen in this decorator besides selecting functionality based on the base class.

This also allows for expansion of functionality beyond current class definitions
`@fastapi.define_class(**other_arg)`
